cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java

```diff
@@ -0,0 +1,555 @@
+      super.visit(version, access, name, signature, superName, interfaces);
+      interfaceClass = Modifier.isInterface(access);
+    }
+
+    @Override
+    public FieldVisitor visitField(int access, final String fieldName, String desc, String signature, Object value) {
+      fieldDetails.put(fieldName, Type.getType(desc));
+      return super.visitField(access, fieldName, desc, signature, value);
+    }
+
+    @Override
+    public MethodVisitor visitMethod(final int access, final String methodName, final String methodDesc,
+                                     final String signature, final String[] exceptions) {
+      MethodVisitor mv = super.visitMethod(access, methodName, methodDesc, signature, exceptions);
+
+      // No need to visit the class to look for AuthEnforce annotation if this class is an interface or the method is
+      // static.
+      if (interfaceClass || Modifier.isStatic(access)) {
+        return mv;
+      }
+
+      // Visit the annotations of the method to determine if it has AuthEnforce annotation. If it does then collect
+      // AuthEnforce annotation details and also sets a boolean flag hasEnforce which is used later in
+      // visitParameterAnnotation to visit the annotations on parameters of this method only if it had AuthEnforce
+      // annotation.
+      return new MethodVisitor(Opcodes.ASM5, mv) {
+
+        final Map<Integer, AnnotationNode> parameterAnnotationNode = new HashMap<>();
+        boolean hasAuthEnforce;
+        AnnotationNode authEnforceAnnotationNode;
+
+        @Override
+        public AnnotationVisitor visitAnnotation(String desc, boolean visible) {
+          AnnotationVisitor annotationVisitor = super.visitAnnotation(desc, visible);
+
+          // If the annotation is not visible or its not AuthEnforce just skip
+          if (!(visible && AuthEnforce.class.getName().equals(Type.getType(desc).getClassName()))) {
+            return annotationVisitor;
+          }
+          hasAuthEnforce = true;
+          authEnforceAnnotationNode = new AnnotationNode(desc);
+          return authEnforceAnnotationNode;
+        }
+
+        @Override
+        public AnnotationVisitor visitParameterAnnotation(final int parameter, String desc, boolean visible) {
+          AnnotationVisitor annotationVisitor = super.visitParameterAnnotation(parameter, desc, visible);
+          if (!(hasAuthEnforce && visible && Name.class.getName().equals(Type.getType(desc).getClassName()))) {
+            return annotationVisitor;
+          }
+          // Since AuthEnforce annotation was used on this method then look for parameters with named annotation and
+          // store their position.
+          // TODO: Should also pick up names from QueryParam and PathParam annotations here as needed later
+          AnnotationNode annotationNode = new AnnotationNode(desc);
+          // store the parameter position and its annotation detail
+          parameterAnnotationNode.put(parameter, annotationNode);
```

Comment:
What if the parameter has more than one annotation? E.g.

```java
void fooBar(@PathParam("id") @DefaultValue("xyz") String id)
```

Also, it might contains multiple annotations that are both recognized as "name" by the rewriter (@Name("n1") @PathParam("id") String id). This mean we should have a precedence for those annotations as well.